workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt

```diff
@@ -0,0 +1,180 @@
+}
+
+/**
+ * @param enforceMainThread Allows disabling the main thread check for testing.
+ */
+@OptIn(WorkflowUiExperimentalApi::class)
+@VisibleForTesting(otherwise = PRIVATE)
+internal class RealWorkflowLifecycleOwner(
+  private val view: View,
+  private val findParentLifecycle: () -> Lifecycle?,
+  enforceMainThread: Boolean = true,
+) : WorkflowLifecycleOwner,
+  LifecycleOwner,
+  OnAttachStateChangeListener,
+  LifecycleEventObserver {
+
+  private val localLifecycle =
+    if (enforceMainThread) LifecycleRegistry(this) else createUnsafe(this)
+  private var parentLifecycle: Lifecycle? = null
+  private var destroyOnDetach = false
+
+  override fun onViewAttachedToWindow(v: View) {
+    val oldLifecycle = parentLifecycle
+    parentLifecycle = findParentLifecycle()
+
+    if (parentLifecycle !== oldLifecycle) {
+      oldLifecycle?.removeObserver(this)
+      parentLifecycle?.addObserver(this)
+    }
+    updateLifecycle(isAttached = true)
+  }
+
+  override fun onViewDetachedFromWindow(v: View) {
+    updateLifecycle(isAttached = false)
+  }
+
+  /** Called when the [parentLifecycle] changes state. */
+  override fun onStateChanged(
+    source: LifecycleOwner,
+    event: Event
+  ) {
+    updateLifecycle()
+  }
+
+  override fun destroyOnDetach() {
+    if (!destroyOnDetach) {
+      destroyOnDetach = true
+      updateLifecycle()
+    }
+  }
+
+  override fun getLifecycle(): Lifecycle = localLifecycle
+
+  /**
+   * @param isAttached Whether the view is [attached][View.isAttachedToWindow]. Must be passed
+   * explicitly when called from the attach/detach callbacks, since the view property's value won't
+   * reflect the new state until after they return.
+   */
+  @VisibleForTesting(otherwise = PRIVATE)
+  internal fun updateLifecycle(isAttached: Boolean = view.isAttachedToWindow) {
+    val parentState = parentLifecycle?.currentState
+    val localState = localLifecycle.currentState
+
+    if (localState == DESTROYED) {
+      // Local destruction is a terminal state.
+      return
+    }
+
+    localLifecycle.currentState = when {
+      // We've been queued for destruction.
+      // Stay attached to the parent's lifecycle until we re-attach, since the parent could be
+      // destroyed while we're detached.
+      destroyOnDetach && !isAttached -> DESTROYED
+      // We may or may not be attached, but we have a parent lifecycle so we just blindly follow it.
+      parentState != null -> parentState
+      // We have no parent but we just attached, so just assume we're resumed.
+      isAttached -> RESUMED
+      // We have no parent and we're detached, so just stay where we are. We don't want to destroy
+      // here since any view in the parent can detach and re-attach us at any time, and that doesn't
+      // mean that we're going away.
+      else -> localState
+    }.also { newState ->
+      if (newState == DESTROYED) {
+        // We're now in a terminal DESTROY state. Be a good citizen and make sure to detach from
```

Comment:
Are we not able to 'push' the destroy to our observers in the case that the view hasn't been attached? I don't see why that is a necessary step.

Reply:
That will happen, but it's not related to this code so I'm not sure I understand what you mean. If the lifecycle has listeners when this code is executed, they will be notified of the destroyed event when we set the value on the `LifecycleRegistry`.

This is just ensuring that we clear our own listener of our parent when we no longer care to listen.

Reply:
Sorry ya my comment was confusing - It was related to your previous kdoc note of making sure we *attach* the view. I was commenting here because I thought this was the place we handled destorying whether ot not the view was ever attached so I didn't see a reason why we needed to do what you cautioned we must do (up on line 50)